libminifi/src/controllers/SSLContextService.cpp

```diff
@@ -128,16 +149,234 @@ bool SSLContextService::configure_ssl_context(SSL_CTX *ctx) {
+    logger_->log_debug("Skipping client certificate %s because its extended key usage set does not contain all usages specified in %s",
+                       x509_cert->name, Configuration::nifi_security_windows_client_cert_key_usage);
+    return false;
+  }
+
+  if (SSL_CTX_use_certificate(ctx, x509_cert.get()) != 1) {
+    logger_->log_error("Failed to set certificate from %s, %s", x509_cert->name, getLatestOpenSSLErrorString);
+    return false;
+  }
+
+  if (SSL_CTX_use_PrivateKey(ctx, private_key.get()) != 1) {
+    logger_->log_error("Failed to use private key %s, %s", x509_cert->name, getLatestOpenSSLErrorString());
+    return false;
+  }
+
+  logger_->log_debug("Found client certificate %s", x509_cert->name);
+
+  return true;
+}
+#endif  // WIN32
+
+bool SSLContextService::addServerCertificatesFromSystemStoreToSSLContext(SSL_CTX* ctx) const {
+#ifdef WIN32
+  utils::tls::WindowsCertStoreLocation store_location{cert_store_location_};
+  HCERTSTORE hCertStore = CertOpenStore(CERT_STORE_PROV_SYSTEM_A, 0, NULL,
+                                        CERT_STORE_OPEN_EXISTING_FLAG | CERT_STORE_READONLY_FLAG | store_location.getBitfieldValue(),
+                                        server_cert_store_.data());
+  if (!hCertStore) {
+    logger_->log_error("Could not open system certificate store %s/%s (server certificates)", cert_store_location_, server_cert_store_);
+    return false;
+  }
+  const auto store_close = gsl::finally([hCertStore](){ CertCloseStore(hCertStore, 0); });
+
+  X509_STORE* ssl_store = SSL_CTX_get_cert_store(ctx);
+  if (!ssl_store) {
+    logger_->log_error("Could not get handle to SSL certificate store");
+    return false;
+  }
+
+  logger_->log_debug("Adding server certificates from system store %s/%s", cert_store_location_, server_cert_store_);
+
+  PCCERT_CONTEXT pCertContext = nullptr;
+  while (pCertContext = CertEnumCertificatesInStore(hCertStore, pCertContext)) {
+    addServerCertificateToSSLStore(ssl_store, pCertContext);
+  }
+
+  return true;
+#else
+  SSL_CTX_set_default_verify_paths(ctx);
+  return true;
+#endif  // WIN32
+}
+
+#ifdef WIN32
+void SSLContextService::addServerCertificateToSSLStore(X509_STORE* ssl_store, PCCERT_CONTEXT certificate) const {
+  utils::tls::X509_unique_ptr x509_cert = utils::tls::convertWindowsCertificate(certificate);
+  if (!x509_cert) {
+    logger_->log_error("Failed to convert system store server certificate to X.509 format");
+    return;
+  }
+
+  int success = X509_STORE_add_cert(ssl_store, x509_cert.get());
+  if (success == 1) {
+    logger_->log_debug("Added server certificate %s from the system store to the SSL store", x509_cert->name);
+  } else if (ERR_peek_last_error() == X509_R_CERT_ALREADY_IN_HASH_TABLE) {
+    logger_->log_debug("Ignoring duplicate server certificate %s", x509_cert->name);
+  } else {
+    logger_->log_error("Failed to add server certificate %s to the SSL store; error: %s", x509_cert->name, getLatestOpenSSLErrorString());
+  }
```

Comment:
I tested the latest change, but unfortunately it does not change the logging. The `X509_R_CERT_ALREADY_IN_HASH_TABLE` we are trying to check has the value `101` but after I log the `ERR_peek_last_error` it returns `201322597` so it is not the error code we were looking for. As checking further the error code in binary is `1011 1111 1111 1111 0000 0110 0101` where the last byte's value is actually `101`.

According to the [OpenSSL documentation](https://www.openssl.org/docs/man1.0.2/man3/ERR_GET_FUNC.html) the returned value of `ERR_peek_last_error` consists of a library number, function code and reason code and there are separate macros defined to retrieve needs. We need to use `ERR_GET_REASON` to retrieve the error code.

I tested the following change and it seems to be working.

```suggestion
  } else {
    auto err = ERR_peek_last_error();
    if (ERR_GET_REASON(err) == X509_R_CERT_ALREADY_IN_HASH_TABLE) {
      logger_->log_debug("Ignoring duplicate server certificate %s", x509_cert->name);
    } else {
      logger_->log_error("Failed to add server certificate %s to the SSL store; error: %s", x509_cert->name, getLatestOpenSSLErrorString());
    }
  }
```

Reply:
Thanks for the fix!  I have rewritten it slightly to avoid nested `if` statements.  Can you retest it, please?

Reply:
I retested it, it's working